File: frontend/src/api/studyPlanner.ts
import { fetchApi } from '../lib/api';

export interface StudyPlan {
  id: number;
  user_id: number;
  title: string;
  exam_name: string;
  exam_date: string;
  available_hours_per_day: number;
  preferred_study_time: string;
  target_score_percentage: number;
  weak_subjects: string[];
  strong_subjects: string[];
  status: 'Active' | 'Completed' | 'Archived';
  created_at: string;
}

export interface StudyPlanInput {
  title: string;
  exam_name: string;
  exam_date: string;
  available_hours_per_day: number;
  preferred_study_time: string;
  target_score_percentage: number;
  weak_subjects: string[];
  strong_subjects: string[];
}

export interface StudyTask {
  id: number;
  plan_id: number;
  user_id: number;
  subject_code: string;
  subject_name: string;
  title: string;
  description?: string;
  task_type: 'Study' | 'Revision' | 'Assignment' | 'Quiz Practice' | 'PYQP Analysis';
  scheduled_date: string;
  duration_minutes: number;
  priority: 'High' | 'Medium' | 'Low';
  status: 'Pending' | 'In Progress' | 'Completed';
  completed_at?: string;
  linked_module_type?: string;
  linked_module_id?: number;
  created_at: string;
}

export interface AiSuggestion {
  id: string;
  title: string;
  description: string;
  subject_name: string;
  suggestion_type: 'assignment' | 'attendance' | 'quiz' | 'pyqp' | 'document';
  action_label: string;
  priority: 'High' | 'Medium' | 'Low';
  module_link?: string;
}

export interface StudyAnalytics {
  days_to_exam: number;
  daily_progress_percentage: number;
  weekly_progress_percentage: number;
  monthly_progress_percentage: number;
  total_study_hours_allocated: number;
  total_study_hours_completed: number;
  completed_topics_count: number;
  remaining_topics_count: number;
  weak_subjects_status: {
    subject_name: string;
    mastery_percentage: number;
    tasks_remaining: number;
  }[];
}

export interface StudyReminder {
  id: number;
  title: string;
  reminder_type: 'Study Time' | 'Assignment Due' | 'Quiz Reminder' | 'Exam Reminder';
  scheduled_time: string;
  is_read: boolean;
}

export const getStudyPlans = async (): Promise<StudyPlan[]> => {
  return fetchApi('/study-planner/plans');
};

export const generateStudyPlan = async (input: StudyPlanInput): Promise<StudyPlan> => {
  return fetchApi('/study-planner/generate', {
    method: 'POST',
    body: JSON.stringify(input),
  });
};

export const deleteStudyPlan = async (id: number): Promise<{ message: string; id: number }> => {
  return fetchApi(`/study-planner/plans/${id}`, {
    method: 'DELETE',
  });
};

export const getStudyTasks = async (params?: {
  date_filter?: 'today' | 'tomorrow' | 'this_week' | 'this_month';
  task_status?: 'Pending' | 'In Progress' | 'Completed';
  subject?: string;
  search?: string;
}): Promise<StudyTask[]> => {
  const searchParams = new URLSearchParams();
  if (params?.date_filter) searchParams.append('date_filter', params.date_filter);
  if (params?.task_status) searchParams.append('task_status', params.task_status);
  if (params?.subject) searchParams.append('subject', params.subject);
  if (params?.search) searchParams.append('search', params.search);

  const queryString = searchParams.toString();
  return fetchApi(`/study-planner/tasks${queryString ? `?${queryString}` : ''}`);
};

export const updateTaskStatus = async (
  id: number,
  status: 'Pending' | 'In Progress' | 'Completed'
): Promise<StudyTask> => {
  return fetchApi(`/study-planner/tasks/${id}/status`, {
    method: 'PATCH',
    body: JSON.stringify({ status }),
  });
};

export const createCustomTask = async (task: {
  plan_id: number;
  subject_code: string;
  subject_name: string;
  title: string;
  description?: string;
  task_type: string;
  scheduled_date: string;
  duration_minutes: number;
  priority: string;
}): Promise<StudyTask> => {
  return fetchApi('/study-planner/tasks', {
    method: 'POST',
    body: JSON.stringify(task),
  });
};

export const deleteStudyTask = async (id: number): Promise<{ message: string; id: number }> => {
  return fetchApi(`/study-planner/tasks/${id}`, {
    method: 'DELETE',
  });
};

export const getAiStudySuggestions = async (): Promise<AiSuggestion[]> => {
  return fetchApi('/study-planner/suggestions');
};

export const getStudyAnalytics = async (): Promise<StudyAnalytics> => {
  return fetchApi('/study-planner/analytics');
};

export const getStudyReminders = async (): Promise<StudyReminder[]> => {
  return fetchApi('/study-planner/reminders');
};

File: backend/app/models/study_planner.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Boolean, Date, Float
from sqlalchemy.sql import func
from app.database.base import Base

class StudyPlanModel(Base):
    __tablename__ = "study_plans"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    title = Column(String(255), nullable=False)
    exam_name = Column(String(255), nullable=False)
    exam_date = Column(Date, nullable=False)
    available_hours_per_day = Column(Float, default=4.0)
    preferred_study_time = Column(String(50), default="Evening") # Morning, Afternoon, Evening, Night
    target_score_percentage = Column(Integer, default=90)
    weak_subjects = Column(Text, nullable=True) # JSON array
    strong_subjects = Column(Text, nullable=True) # JSON array
    status = Column(String(50), default="Active", index=True) # Active, Completed, Archived
    created_at = Column(DateTime(timezone=True), server_default=func.now())

class StudyTaskModel(Base):
    __tablename__ = "study_tasks"

    id = Column(Integer, primary_key=True)
    plan_id = Column(Integer, ForeignKey("study_plans.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    subject_code = Column(String(50), nullable=False, index=True)
    subject_name = Column(String(150), nullable=False, index=True)
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    task_type = Column(String(50), default="Study", index=True) # Study, Revision, Assignment, Quiz Practice, PYQP Analysis
    scheduled_date = Column(Date, nullable=False, index=True)
    duration_minutes = Column(Integer, default=60)
    priority = Column(String(20), default="Medium") # High, Medium, Low
    status = Column(String(50), default="Pending", index=True) # Pending, In Progress, Completed
    completed_at = Column(DateTime(timezone=True), nullable=True)
    linked_module_type = Column(String(50), nullable=True) # assignment, quiz, pyqp, document, attendance
    linked_module_id = Column(Integer, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

class StudyReminderModel(Base):
    __tablename__ = "study_reminders"

    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    plan_id = Column(Integer, ForeignKey("study_plans.id", ondelete="CASCADE"), nullable=True)
    title = Column(String(255), nullable=False)
    reminder_type = Column(String(50), default="Study Time") # Study Time, Assignment Due, Quiz Reminder, Exam Reminder
    scheduled_time = Column(DateTime(timezone=True), nullable=False)
    is_read = Column(Boolean, default=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
